activerecord/test/cases/relation/merging_test.rb

```diff
@@ -213,23 +213,55 @@ def test_relation_merging_with_locks
     assert_predicate devs, :locked?
   end
 
+  def test_relation_merging_with_select
+    comments = Comment.select(:id, :post_id).merge(Comment.select(:id, :post_id))
+    assert_equal Comment.pluck(:id).sort, comments.to_a.map(&:id).sort
+    assert_equal Comment.pluck(:post_id).sort, comments.to_a.map(&:post_id).sort
+    assert_equal [:id, :post_id], comments.select_values
+  end
+
+  def test_relation_merging_with_select_and_sti
+    comments = Comment.select(:id, :post_id).merge(SpecialComment.select(:id, :post_id))
+    assert_equal SpecialComment.pluck(:id).sort, comments.to_a.map(&:id).sort
+    assert_equal SpecialComment.pluck(:post_id).sort, comments.to_a.map(&:post_id).sort
+    assert_equal [:id, :post_id], comments.select_values
+  end
+
   def test_relation_merging_with_preload
     [Post.all.merge(Post.preload(:author)), Post.preload(:author).merge(Post.all)].each do |posts|
       assert_queries_count(2) { assert posts.first.author }
     end
   end
 
+  def test_relation_merging_with_preload_and_sti
+    [Post.all.merge(SpecialPost.preload(:author)), Post.preload(:author).merge(Post.all)].each do |posts|
+      assert_equal [:author], posts.preload_values
+      assert_queries_count(2) { assert posts.first.author }
+    end
+  end
```

Comment:
This actually raises without the patch:
![Screenshot 2025-01-15 at 6 13 22 pm](https://github.com/user-attachments/assets/f8368c74-e7da-42a3-aa10-9a86af0b3351)